### ISSUE TITLE
Rollup didn't finish over 30 minutes.

### ISSUE THREAD

**Issue by mysticatea** [rollup/rollup#2212]
Repro: https://github.com/mysticatea/rollup-issue

I tried to rollup the typescript language service (this will be run in a WebWorker) in [MonacoEditor] library, but it didn't finish in over 30 minutes.

I suspected it might dive into infinite looping because it was using a CPU core heavily with no warnings.

The configuration file is simple:

```js
export default {
    input: require.resolve("monaco-editor/esm/vs/language/typescript/ts.worker.js"),
    output: {
        file: "dist/ts.worker.js",
        format: "iife",
        name: "tsWorker",
    }
}
```

Also, I tried some options:

- `perf: true`, but it didn't affect -- console output is not changed.
- `treeshake: false`, but it didn't affect -- rollup still doesn't finish.

[MonacoEditor]: https://github.com/Microsoft/monaco-editor

**Comment by mysticatea** [comment]
I'm sorry, I should ignore some files to build it. (https://github.com/Microsoft/monaco-editor-samples/blob/master/browser-esm-webpack/webpack.config.js#L26)

**Comment by mysticatea** [comment]
The ignoring seems to not related to this issue (those are used in `require()`).

**Comment by mysticatea** [comment]
This might be the same problem as #2210.

**Comment by lukastaegert** [comment]
Ok, it seems there are still >100000 lines of code in this...

**Comment by lukastaegert** [comment]
Ok, I managed to track the issue down to a few thousand lines of code in the TS service that are exceptionally slow when bundled in Rollup (but not frozen in fact—its 6000 lines that take about 40s). Using Rollup's own performance timers showed that it was indeed not the tree-shaking itself but the graph analysis that was slow (65ms vs >40000ms).

Firing up Chrome dev tools and creating an execution profile hints to it that it is NOT the literal value tracking but indeed the "improved" reassignment logic that is causing issues. More specifically the reassigment logic in combination with the return expression tracking. At the moment, all return values of a function are tracked. That means when a nested value of an object returned by a call is marked as reassigned, the corresponding paths of all return expressions will be marked as reassigned. If this is tracked along several nested functions, you can easily see where the exponential performance degradation could come from.

I do not have a lot of time on my hands today but I will try to post a fix soon. The approach will probably be to treat functions with more than one return statement as returning "unknown" + appropriate deoptimizations. This would be similar to how a variable with more than one assignment is treated as "unknown" at the moment.

**Comment by lukastaegert** [comment]
Fix released as 0.59.4
